File: src/falcon_core/math/discrete_spaces/discretizer.py
from __future__ import annotations
from typing import Any, List, Dict, Tuple, Optional
from falcon_core._capi.discretizer import Discretizer as _CDiscretizer
from falcon_core.math.domains.domain import Domain

class Discretizer:
    """Python wrapper for Discretizer."""

    def __init__(self, c_obj):
        self._c = c_obj

    @classmethod
    def _from_capi(cls, c_obj):
        if c_obj is None:
            return None
        return cls(c_obj)

    @classmethod
    def new_cartesian_discretizer(cls, delta: Any) -> Discretizer:
        return cls(_CDiscretizer.new_cartesian_discretizer(delta))

    @classmethod
    def new_polar_discretizer(cls, delta: Any) -> Discretizer:
        return cls(_CDiscretizer.new_polar_discretizer(delta))

    @classmethod
    def from_json(cls, json: str) -> Discretizer:
        return cls(_CDiscretizer.from_json(json))

    def delta(self, ) -> None:
        ret = self._c.delta()
        return ret

    def set_delta(self, delta: Any) -> None:
        ret = self._c.set_delta(delta)
        return ret

    def domain(self, ) -> Domain:
        ret = self._c.domain()
        if ret is None: return None
        return Domain._from_capi(ret)

    def is_cartesian(self, ) -> None:
        ret = self._c.is_cartesian()
        return ret

    def is_polar(self, ) -> None:
        ret = self._c.is_polar()
        return ret

    def equal(self, b: Discretizer) -> None:
        ret = self._c.equal(b._c)
        return ret

    def not_equal(self, b: Discretizer) -> None:
        ret = self._c.not_equal(b._c)
        return ret

    def __eq__(self, other):
        """Operator overload for =="""
        if not isinstance(other, Discretizer):
            return NotImplemented
        return self.equal(other)

    def __ne__(self, other):
        """Operator overload for !="""
        if not isinstance(other, Discretizer):
            return NotImplemented
        return self.not_equal(other)

File: ocaml/_opam/.opam-switch/sources/falcon_core/src/falcon_core/math/discrete_spaces/discretizer.ml
open Ctypes
open Capi_bindings
open ErrorHandling

open Falcon_core.Math.Domains

class c_discretizer (h : unit ptr) = object(self)
  val raw_val = h
  method raw = raw_val
  initializer Gc.finalise (fun _ ->
    Capi_bindings.discretizer_destroy raw_val;
    ErrorHandling.raise_if_error ()
  ) self
end

module Discretizer = struct
  type t = c_discretizer

  let copy (handle : t) : t =
    ErrorHandling.read handle (fun () ->
      let ptr = Capi_bindings.discretizer_copy handle#raw in
      ErrorHandling.raise_if_error ();
      new c_discretizer ptr
    )

  let fromjson (json : string) : t =
    let ptr = Capi_bindings.discretizer_from_json_string (Capi_bindings.string_wrap json) in
    ErrorHandling.raise_if_error ();
    new c_discretizer ptr

  let cartesianDiscretizer (delta : float) : t =
    let ptr = Capi_bindings.discretizer_create_cartesian_discretizer delta in
    ErrorHandling.raise_if_error ();
    new c_discretizer ptr

  let polarDiscretizer (delta : float) : t =
    let ptr = Capi_bindings.discretizer_create_polar_discretizer delta in
    ErrorHandling.raise_if_error ();
    new c_discretizer ptr

  let equal (handle : t) (other : t) : bool =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.discretizer_equal handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let notEqual (handle : t) (other : t) : bool =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.discretizer_not_equal handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let toJsonString (handle : t) : string =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.discretizer_to_json_string handle#raw in
      ErrorHandling.raise_if_error ();
      Capi_bindings.string_to_ocaml result
    )

  let delta (handle : t) : float =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.discretizer_delta handle#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let setDelta (handle : t) (delta : float) : unit =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.discretizer_set_delta handle#raw delta in
      ErrorHandling.raise_if_error ();
      result
    )

  let domain (handle : t) : Domain.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.discretizer_domain handle#raw in
      ErrorHandling.raise_if_error ();
      new c_domain result
    )

  let isCartesian (handle : t) : bool =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.discretizer_is_cartesian handle#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let isPolar (handle : t) : bool =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.discretizer_is_polar handle#raw in
      ErrorHandling.raise_if_error ();
      result
    )

end